auto_nag/scripts/inactive_ni_pending.py

```diff
@@ -0,0 +1,88 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+
+from auto_nag import utils
+from auto_nag.bzcleaner import BzCleaner
+from auto_nag.user_activity import UserActivity
+
+
+class InactiveNeedinfoPending(BzCleaner):
+    def __init__(self):
+        super(InactiveNeedinfoPending, self).__init__()
+
+    def description(self):
+        return "Bugs with needinfo pending on inactive people"
+
+    def columns(self):
+        return ["id", "summary", "inactive_ni", "inactive_ni_count"]
+
+    def get_bugs(self, *args, **kwargs):
+        bugs = super().get_bugs(*args, **kwargs)
+        self.handle_inactive_requestee(bugs)
+
+        # Resolving https://github.com/mozilla/relman-auto-nag/issues/1300 should clean this
+        # including improve the wording in the template (i.e., "See the search query on Bugzilla").
+        self.query_url = utils.get_bz_search_url({"bug_id": ",".join(bugs.keys())})
+
+        return bugs
+
+    def handle_inactive_requestee(self, bugs):
+        requestee_bugs = {}
+        for bugid, bug in bugs.items():
+            for flag in bug["needinfo_flags"]:
+                if flag["requestee"] not in requestee_bugs:
+                    requestee_bugs[flag["requestee"]] = [bugid]
+                else:
+                    requestee_bugs[flag["requestee"]].append(bugid)
+
+        user_activity = UserActivity()
+        inactive_users = user_activity.check_users(requestee_bugs.keys())
+        selected_bugs = []
+        for requestee, bugids in requestee_bugs.items():
+            if requestee in inactive_users:
+                selected_bugs.extend(bugids)
+
+        for bugid in set(bugs.keys()) - set(selected_bugs):
+            del bugs[bugid]
+
+        # Add the requestee info for the email template
+        for bug in bugs.values():
+            bug["inactive_ni"] = [
+                (
+                    flag["requestee"],
+                    user_activity.get_string_status(inactive_users[flag["requestee"]]),
+                )
+                for flag in bug["needinfo_flags"]
+                if flag["requestee"] in inactive_users
+            ]
+            bug["inactive_ni_count"] = len(bug["inactive_ni"])
+
+    def handle_bug(self, bug, data):
+        bugid = str(bug["id"])
+        data[bugid] = {
+            "needinfo_flags": [
+                flag
+                for flag in bug["flags"]
+                if "requestee" in flag and flag["name"] == "needinfo"
+            ]
+        }
+
+        return bug
+
+    def get_bz_params(self, date):
+        fields = ["flags"]
+        params = {
+            "include_fields": fields,
+            "resolution": "---",
+            "f6": "flagtypes.name",
+            "o6": "equals",
+            "v6": "needinfo?",
+        }
```

Comment:
We could exclude bugs that have been changed recently (e.g. last month), we can assume the bug is being actively worked on even if the needinfo hasn't been answered yet.

Reply:
Maybe there is some other filter we could use to reduce the number here

Reply:
> We could exclude bugs that have been changed recently (e.g. last month), we can assume the bug is being actively worked on even if the needinfo hasn't been answered yet.

I think such a warning is more important for active bugs than old ones. Needinfo on inactive users could be blocking a bug that is being actively worked on.

Reply:
> Maybe there is some other filter we could use to reduce the number here

The number of bugs is currently small, and I hope it will be even smaller after deploying the tool (as an impact).

Reply:
> Maybe there is some other filter we could use to reduce the number here

We could exclude very old bugs (e.g. created 5 years ago).

Reply:
> My reasoning was that if a bug is actively being worked on, people won't needinfo somebody who is actually inactive (especially if their account is disabled). Usually the needinfos on inactive people are old needinfos that never got answered by the reporter (maybe a user who reported a bug and never added the information required to move forward) or by a developer (maybe somebody who left Mozilla since then).


I agree, valid point. In the same time if for any reason a needinfo requested on an inactive user, it could be helpful to get a warning from the bot.



> We'd lose the possibility of cleaning the backlog. How many results does the query return if we limit to 5 years? Another possibility could be to run weekly on bugs from the past year, and monthly on all bugs.


| | Without filters  | Created in last 5 years | Not changed in last month|
| :-- |  :--: |  :--: | :--: |
| The initial query | 5,345 bugs | 4,275 bugs| 4,329 bugs |
| With ni on inactive users |1,415 bugs | 1,077 bugs|  1,322 bugs  |

Reply:
All right, let's do this way then:
- run monthly on all bugs with no filters (super slow);
- run weekly only on a subset of bugs (e.g. opened/changed since a month before, or some other filter).

Do you agree?

Reply:
I agree. This way, we don't lose the possibility of clearing the backlog, and at the same time, we help with the recent bugs.

For bugs opened in the last month, we have the following:
- The initial query: 1,010 bugs
- With ni on inactive users: 93 bugs
- Executed in less than 6 mins

Running weekly with this filter is good for now. Later, the number of bugs will be lower (as an impact of this tool), then we could even run it daily.